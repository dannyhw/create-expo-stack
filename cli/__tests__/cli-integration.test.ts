import Bun from 'bun';

import { version } from '../package.json';

import { test, expect } from 'bun:test';
import * as path from 'node:path';

type InputFlag = `--${string}`;

const cli = async (inputs: string[]) => {
  const pathToFile = `${path.join(__dirname, '../', 'bin', 'create-expo-stack.js')}`;

  console.log('running', `bun ${pathToFile} ${inputs.join(' ')}`);

  const { stdout, exitCode, success, stderr } = Bun.spawnSync(['bun', pathToFile, ...inputs]);

  if (!success || exitCode !== 0) {
    const stdoutStr = stdout.toString();
    console.log('failed command', `bun ${pathToFile} ${inputs.join(' ')}`);
    console.log('stderr: ', stderr.toString());
    console.log('stdout: ', stdoutStr);
    throw new Error(stderr.toString());
  }

  return stdout.toString();
};

// we can generate combinations soon.
const generateProject = async ({
  projectName = 'myTestProject',
  flags
}: {
  projectName?: string;
  flags: Array<InputFlag>;
}) => {
  return cli([projectName, ...flags]);
};

// Run tests for each package manager
const packageManagers = process.env.ALL_PACKAGE_MANAGERS
  ? ([`npm`, `yarn`, `pnpm`, `bun`] as const)
  : (['npm'] as const);

test(`outputs version`, async () => {
  const output = await cli([`--version`]);

  expect(output).toContain(version);
});

test(`outputs help`, async () => {
  const output = await cli([`--help`]);

  expect(output).toContain(`Info`);
});

const popularCombinations = [
  ['--expo-router', '--nativewind'],
  ['--expo-router', '--stylesheet'],
  ['--expo-router', '--tabs', '--nativewind'],
  ['--expo-router', '--tabs', '--stylesheet'],
  ['--expo-router', '--drawer+tabs', '--nativewind'],
  ['--expo-router', '--drawer+tabs', '--stylesheet']
] as const;

// let i = 0;

const projectName = `myTestProject`;

for (const packageManager of packageManagers) {
  const packageManagerFlag = `--${packageManager}` as const;
  for (const flags of popularCombinations) {
    // const projectName = `myTestProject-${i++}`;

    const finalFlags = [...flags, packageManagerFlag, '--overwrite' as const];

    test(`generates a project with ${finalFlags.join(', ')}`, async () => {
      const output = await generateProject({
        projectName: projectName,
        flags: finalFlags
      });

      expect(output).toContain(packageManager);

      expect(output).toContain('Installing dependencies');
    });
  }
}

// i18next
test(`generates a default project with i18n`, async () => {
  const output = await generateProject({
    projectName: 'myTestProject',
    flags: ['--default', `--i18next`, `--bun`]
  });

  expect(output).toContain('--i18next');
});
