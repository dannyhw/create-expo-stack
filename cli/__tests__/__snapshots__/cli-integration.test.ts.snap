// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`generates a project with --expo-router, --nativewind, --npm, --overwrite 1`] = `
{
  "dependencies": {
    "@expo/vector-icons": "",
    "@react-navigation/native": "",
    "expo": "",
    "expo-constants": "",
    "expo-linking": "",
    "expo-router": "",
    "expo-status-bar": "",
    "expo-system-ui": "",
    "expo-web-browser": "",
    "nativewind": "",
    "react": "",
    "react-dom": "",
    "react-native": "",
    "react-native-gesture-handler": "",
    "react-native-reanimated": "",
    "react-native-safe-area-context": "",
    "react-native-screens": "",
    "react-native-web": "",
  },
  "devDependencies": {
    "@babel/core": "",
    "@types/react": "",
    "@typescript-eslint/eslint-plugin": "",
    "@typescript-eslint/parser": "",
    "eslint": "",
    "eslint-config-universe": "",
    "prettier": "",
    "prettier-plugin-tailwindcss": "",
    "tailwindcss": "",
    "typescript": "",
  },
  "eslintConfig": {
    "extends": "universe/native",
    "root": true,
  },
  "expo": {
    "install": {
      "exclude": [
        "react-native-safe-area-context",
      ],
    },
  },
  "main": "expo-router/entry",
  "name": "myTestProject",
  "private": true,
  "scripts": {
    "android": "expo start --android",
    "format": "eslint "**/*.{js,jsx,ts,tsx}" --fix && prettier "**/*.{js,jsx,ts,tsx,json}" --write",
    "ios": "expo start --ios",
    "lint": "eslint "**/*.{js,jsx,ts,tsx}" && prettier -c "**/*.{js,jsx,ts,tsx,json}"",
    "prebuild": "expo prebuild",
    "start": "expo start",
    "web": "expo start --web",
  },
  "version": "1.0.0",
}
`;

exports[`generates a project with --expo-router, --nativewind, --npm, --overwrite 2`] = `
{
  "cesVersion": undefined,
  "flags": {
    "eas": false,
    "importAlias": true,
    "noGit": false,
    "noInstall": false,
    "overwrite": true,
    "packageManager": "npm",
  },
  "os": {},
  "packageManager": {
    "type": "npm",
    "version": undefined,
  },
  "packages": [
    {
      "name": "expo-router",
      "options": {
        "type": "stack",
      },
      "type": "navigation",
    },
    {
      "name": "nativewind",
      "type": "styling",
    },
  ],
  "projectName": "myTestProject",
}
`;

exports[`generates a project with --expo-router, --stylesheet, --npm, --overwrite 1`] = `
{
  "dependencies": {
    "@expo/vector-icons": "",
    "@react-navigation/native": "",
    "expo": "",
    "expo-constants": "",
    "expo-linking": "",
    "expo-router": "",
    "expo-status-bar": "",
    "expo-system-ui": "",
    "expo-web-browser": "",
    "nativewind": "",
    "react": "",
    "react-dom": "",
    "react-native": "",
    "react-native-gesture-handler": "",
    "react-native-reanimated": "",
    "react-native-safe-area-context": "",
    "react-native-screens": "",
    "react-native-web": "",
  },
  "devDependencies": {
    "@babel/core": "",
    "@types/react": "",
    "@typescript-eslint/eslint-plugin": "",
    "@typescript-eslint/parser": "",
    "eslint": "",
    "eslint-config-universe": "",
    "prettier": "",
    "prettier-plugin-tailwindcss": "",
    "tailwindcss": "",
    "typescript": "",
  },
  "eslintConfig": {
    "extends": "universe/native",
    "root": true,
  },
  "expo": {
    "install": {
      "exclude": [
        "react-native-safe-area-context",
      ],
    },
  },
  "main": "expo-router/entry",
  "name": "myTestProject",
  "private": true,
  "scripts": {
    "android": "expo start --android",
    "format": "eslint "**/*.{js,jsx,ts,tsx}" --fix && prettier "**/*.{js,jsx,ts,tsx,json}" --write",
    "ios": "expo start --ios",
    "lint": "eslint "**/*.{js,jsx,ts,tsx}" && prettier -c "**/*.{js,jsx,ts,tsx,json}"",
    "prebuild": "expo prebuild",
    "start": "expo start",
    "web": "expo start --web",
  },
  "version": "1.0.0",
}
`;

exports[`generates a project with --expo-router, --stylesheet, --npm, --overwrite 2`] = `
{
  "cesVersion": undefined,
  "flags": {
    "eas": false,
    "importAlias": true,
    "noGit": false,
    "noInstall": false,
    "overwrite": true,
    "packageManager": "npm",
  },
  "os": {},
  "packageManager": {
    "type": "npm",
    "version": undefined,
  },
  "packages": [
    {
      "name": "expo-router",
      "options": {
        "type": "stack",
      },
      "type": "navigation",
    },
    {
      "name": "nativewind",
      "type": "styling",
    },
  ],
  "projectName": "myTestProject",
}
`;

exports[`generates a project with --expo-router, --tabs, --nativewind, --npm, --overwrite 1`] = `
{
  "dependencies": {
    "@expo/vector-icons": "",
    "@react-navigation/native": "",
    "expo": "",
    "expo-constants": "",
    "expo-linking": "",
    "expo-router": "",
    "expo-status-bar": "",
    "expo-system-ui": "",
    "expo-web-browser": "",
    "nativewind": "",
    "react": "",
    "react-dom": "",
    "react-native": "",
    "react-native-gesture-handler": "",
    "react-native-reanimated": "",
    "react-native-safe-area-context": "",
    "react-native-screens": "",
    "react-native-web": "",
  },
  "devDependencies": {
    "@babel/core": "",
    "@types/react": "",
    "@typescript-eslint/eslint-plugin": "",
    "@typescript-eslint/parser": "",
    "eslint": "",
    "eslint-config-universe": "",
    "prettier": "",
    "prettier-plugin-tailwindcss": "",
    "tailwindcss": "",
    "typescript": "",
  },
  "eslintConfig": {
    "extends": "universe/native",
    "root": true,
  },
  "expo": {
    "install": {
      "exclude": [
        "react-native-safe-area-context",
      ],
    },
  },
  "main": "expo-router/entry",
  "name": "myTestProject",
  "private": true,
  "scripts": {
    "android": "expo start --android",
    "format": "eslint "**/*.{js,jsx,ts,tsx}" --fix && prettier "**/*.{js,jsx,ts,tsx,json}" --write",
    "ios": "expo start --ios",
    "lint": "eslint "**/*.{js,jsx,ts,tsx}" && prettier -c "**/*.{js,jsx,ts,tsx,json}"",
    "prebuild": "expo prebuild",
    "start": "expo start",
    "web": "expo start --web",
  },
  "version": "1.0.0",
}
`;

exports[`generates a project with --expo-router, --tabs, --nativewind, --npm, --overwrite 2`] = `
{
  "cesVersion": undefined,
  "flags": {
    "eas": false,
    "importAlias": true,
    "noGit": false,
    "noInstall": false,
    "overwrite": true,
    "packageManager": "npm",
  },
  "os": {},
  "packageManager": {
    "type": "npm",
    "version": undefined,
  },
  "packages": [
    {
      "name": "expo-router",
      "options": {
        "type": "stack",
      },
      "type": "navigation",
    },
    {
      "name": "nativewind",
      "type": "styling",
    },
  ],
  "projectName": "myTestProject",
}
`;

exports[`generates a project with --expo-router, --tabs, --stylesheet, --npm, --overwrite 1`] = `
{
  "dependencies": {
    "@expo/vector-icons": "",
    "@react-navigation/native": "",
    "expo": "",
    "expo-constants": "",
    "expo-linking": "",
    "expo-router": "",
    "expo-status-bar": "",
    "expo-system-ui": "",
    "expo-web-browser": "",
    "nativewind": "",
    "react": "",
    "react-dom": "",
    "react-native": "",
    "react-native-gesture-handler": "",
    "react-native-reanimated": "",
    "react-native-safe-area-context": "",
    "react-native-screens": "",
    "react-native-web": "",
  },
  "devDependencies": {
    "@babel/core": "",
    "@types/react": "",
    "@typescript-eslint/eslint-plugin": "",
    "@typescript-eslint/parser": "",
    "eslint": "",
    "eslint-config-universe": "",
    "prettier": "",
    "prettier-plugin-tailwindcss": "",
    "tailwindcss": "",
    "typescript": "",
  },
  "eslintConfig": {
    "extends": "universe/native",
    "root": true,
  },
  "expo": {
    "install": {
      "exclude": [
        "react-native-safe-area-context",
      ],
    },
  },
  "main": "expo-router/entry",
  "name": "myTestProject",
  "private": true,
  "scripts": {
    "android": "expo start --android",
    "format": "eslint "**/*.{js,jsx,ts,tsx}" --fix && prettier "**/*.{js,jsx,ts,tsx,json}" --write",
    "ios": "expo start --ios",
    "lint": "eslint "**/*.{js,jsx,ts,tsx}" && prettier -c "**/*.{js,jsx,ts,tsx,json}"",
    "prebuild": "expo prebuild",
    "start": "expo start",
    "web": "expo start --web",
  },
  "version": "1.0.0",
}
`;

exports[`generates a project with --expo-router, --tabs, --stylesheet, --npm, --overwrite 2`] = `
{
  "cesVersion": undefined,
  "flags": {
    "eas": false,
    "importAlias": true,
    "noGit": false,
    "noInstall": false,
    "overwrite": true,
    "packageManager": "npm",
  },
  "os": {},
  "packageManager": {
    "type": "npm",
    "version": undefined,
  },
  "packages": [
    {
      "name": "expo-router",
      "options": {
        "type": "stack",
      },
      "type": "navigation",
    },
    {
      "name": "nativewind",
      "type": "styling",
    },
  ],
  "projectName": "myTestProject",
}
`;

exports[`generates a project with --expo-router, --drawer+tabs, --nativewind, --npm, --overwrite 1`] = `
{
  "dependencies": {
    "@expo/vector-icons": "",
    "@react-navigation/native": "",
    "expo": "",
    "expo-constants": "",
    "expo-linking": "",
    "expo-router": "",
    "expo-status-bar": "",
    "expo-system-ui": "",
    "expo-web-browser": "",
    "nativewind": "",
    "react": "",
    "react-dom": "",
    "react-native": "",
    "react-native-gesture-handler": "",
    "react-native-reanimated": "",
    "react-native-safe-area-context": "",
    "react-native-screens": "",
    "react-native-web": "",
  },
  "devDependencies": {
    "@babel/core": "",
    "@types/react": "",
    "@typescript-eslint/eslint-plugin": "",
    "@typescript-eslint/parser": "",
    "eslint": "",
    "eslint-config-universe": "",
    "prettier": "",
    "prettier-plugin-tailwindcss": "",
    "tailwindcss": "",
    "typescript": "",
  },
  "eslintConfig": {
    "extends": "universe/native",
    "root": true,
  },
  "expo": {
    "install": {
      "exclude": [
        "react-native-safe-area-context",
      ],
    },
  },
  "main": "expo-router/entry",
  "name": "myTestProject",
  "private": true,
  "scripts": {
    "android": "expo start --android",
    "format": "eslint "**/*.{js,jsx,ts,tsx}" --fix && prettier "**/*.{js,jsx,ts,tsx,json}" --write",
    "ios": "expo start --ios",
    "lint": "eslint "**/*.{js,jsx,ts,tsx}" && prettier -c "**/*.{js,jsx,ts,tsx,json}"",
    "prebuild": "expo prebuild",
    "start": "expo start",
    "web": "expo start --web",
  },
  "version": "1.0.0",
}
`;

exports[`generates a project with --expo-router, --drawer+tabs, --nativewind, --npm, --overwrite 2`] = `
{
  "cesVersion": undefined,
  "flags": {
    "eas": false,
    "importAlias": true,
    "noGit": false,
    "noInstall": false,
    "overwrite": true,
    "packageManager": "npm",
  },
  "os": {},
  "packageManager": {
    "type": "npm",
    "version": undefined,
  },
  "packages": [
    {
      "name": "expo-router",
      "options": {
        "type": "stack",
      },
      "type": "navigation",
    },
    {
      "name": "nativewind",
      "type": "styling",
    },
  ],
  "projectName": "myTestProject",
}
`;

exports[`generates a project with --expo-router, --drawer+tabs, --stylesheet, --npm, --overwrite 1`] = `
{
  "dependencies": {
    "@expo/vector-icons": "",
    "@react-navigation/native": "",
    "expo": "",
    "expo-constants": "",
    "expo-linking": "",
    "expo-router": "",
    "expo-status-bar": "",
    "expo-system-ui": "",
    "expo-web-browser": "",
    "nativewind": "",
    "react": "",
    "react-dom": "",
    "react-native": "",
    "react-native-gesture-handler": "",
    "react-native-reanimated": "",
    "react-native-safe-area-context": "",
    "react-native-screens": "",
    "react-native-web": "",
  },
  "devDependencies": {
    "@babel/core": "",
    "@types/react": "",
    "@typescript-eslint/eslint-plugin": "",
    "@typescript-eslint/parser": "",
    "eslint": "",
    "eslint-config-universe": "",
    "prettier": "",
    "prettier-plugin-tailwindcss": "",
    "tailwindcss": "",
    "typescript": "",
  },
  "eslintConfig": {
    "extends": "universe/native",
    "root": true,
  },
  "expo": {
    "install": {
      "exclude": [
        "react-native-safe-area-context",
      ],
    },
  },
  "main": "expo-router/entry",
  "name": "myTestProject",
  "private": true,
  "scripts": {
    "android": "expo start --android",
    "format": "eslint "**/*.{js,jsx,ts,tsx}" --fix && prettier "**/*.{js,jsx,ts,tsx,json}" --write",
    "ios": "expo start --ios",
    "lint": "eslint "**/*.{js,jsx,ts,tsx}" && prettier -c "**/*.{js,jsx,ts,tsx,json}"",
    "prebuild": "expo prebuild",
    "start": "expo start",
    "web": "expo start --web",
  },
  "version": "1.0.0",
}
`;

exports[`generates a project with --expo-router, --drawer+tabs, --stylesheet, --npm, --overwrite 2`] = `
{
  "cesVersion": undefined,
  "flags": {
    "eas": false,
    "importAlias": true,
    "noGit": false,
    "noInstall": false,
    "overwrite": true,
    "packageManager": "npm",
  },
  "os": {},
  "packageManager": {
    "type": "npm",
    "version": undefined,
  },
  "packages": [
    {
      "name": "expo-router",
      "options": {
        "type": "stack",
      },
      "type": "navigation",
    },
    {
      "name": "nativewind",
      "type": "styling",
    },
  ],
  "projectName": "myTestProject",
}
`;
